src/ImageSharp/Common/Helpers/SimdUtils.Avx2Intrinsics.cs

```diff
@@ -91,6 +91,190 @@ public static class Avx2Intrinsics
+                ref ReadOnlySpan<byte> channel1,
+                ref ReadOnlySpan<byte> channel2,
+                ref Span<byte> dest)
+            {
+                DebugGuard.IsTrue(channel0.Length == dest.Length, nameof(channel0), "Input spans must be of same length!");
+                DebugGuard.IsTrue(channel1.Length == dest.Length, nameof(channel1), "Input spans must be of same length!");
+                DebugGuard.IsTrue(channel2.Length == dest.Length, nameof(channel2), "Input spans must be of same length!");
+
+                if (Avx2.IsSupported)
+                {
+                    int remainder = ImageMaths.ModuloP2(channel1.Length, Vector256<byte>.Count);
+                    int adjustedCount = channel1.Length - remainder;
+
+                    if (adjustedCount > 0)
+                    {
+                        channel0 = channel0.Slice(adjustedCount);
+                        channel1 = channel1.Slice(adjustedCount);
+                        channel2 = channel2.Slice(adjustedCount);
+                        dest = dest.Slice(adjustedCount);
+
+                        PackBytesToUInt32SaturateChannel4(
+                            channel0,
+                            channel1,
+                            channel2,
+                            dest);
+                    }
+                }
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal static void PackBytesToUInt32SaturateChannel4(
+                ReadOnlySpan<byte> channel0,
+                ReadOnlySpan<byte> channel1,
+                ReadOnlySpan<byte> channel2,
+                Span<byte> dest)
+            {
+                int n = dest.Length / Vector256<byte>.Count;
+
+                ref Vector256<byte> source0Base =
+                    ref Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(channel0));
+                ref Vector256<byte> source1Base =
+                    ref Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(channel1));
+                ref Vector256<byte> source2Base =
+                    ref Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(channel2));
+
+                ref Vector256<byte> destBase = ref Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(dest));
+
+                Vector256<byte> allOnes = Avx2.CompareEqual(Vector256<byte>.Zero, Vector256<byte>.Zero);
+
+                for (int i = 0, j = 0; j < n; i += 1, j += 4)
+                {
+                    Vector256<byte> s0 = Unsafe.Add(ref source0Base, i);
+                    Vector256<byte> s1 = Unsafe.Add(ref source1Base, i);
+                    Vector256<byte> s2 = Unsafe.Add(ref source2Base, i);
+
+                    s0 = Avx2.Permute4x64(s0.AsUInt64(), 0b_11_01_10_00).AsByte();
+                    s1 = Avx2.Permute4x64(s1.AsUInt64(), 0b_11_01_10_00).AsByte();
+                    s2 = Avx2.Permute4x64(s2.AsUInt64(), 0b_11_01_10_00).AsByte();
```

Comment:
Permutes are expensive, so you'll do better to just let the upper/lower lanes stay as is and fix them at the end rather than trying to keep everything in order all the way through.  It may help to write the SSE2 version first and have the AVX2 code match that right up until the end when you deal with the 128-bit lanes being interleaved.

Reply:
:+1: that's actually news to me 😄 , i was always under the impression permutes were only a couple of cycles

Reply:
They're 3 cycles latency and 1 cycle throughput, but they're also only able to be scheduled on a single port, so they can easily be a bottleneck.  I should have said 'relatively expensive', compared to most of the other instructions used here.

A good test to judge the impact would be to write up a benchmark for the SSE2 version of the code to see how the AVX2 version compares to that.  Your goal would, of course, be 2x performance.  You might be surprised what kind of impact all those permutes have in that context :)